### ISSUE TITLE
Issue with installing and Uninstalling

### ISSUE THREAD

**Issue by greenlanegreb** [JMAConsulting/biz.jmaconsulting.grantapplications#168]
Drupal Version 9.1.4
CiviCRM: 5.34.0

On installing - screen goes blank
(can hit back) and comes up as installed
Database Updated Needed - click this it errors with:

[Error: Upgrade biz.jmaconsulting.grantapplications to revision 4301]
DB Error: unknown error
Can be disabled
On attempting to uninstall - screen blank

Thanks and Kind Regards, 

Beanie. 

`#2 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/pear/pear-core-minimal/src/PEAR.php(575): DB_Error->__construct(-1, 16, (Array:2), "ALTER TABLE  `civicrm_grant_app_page` ADD `is_draft` TINYINT(4) DEFAULT NULL ...")
#3 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/pear/pear-core-minimal/src/PEAR.php(223): PEAR->_raiseError(Object(DB_mysqli), NULL, -1, 16, (Array:2), "ALTER TABLE  `civicrm_grant_app_page` ADD `is_draft` TINYINT(4) DEFAULT NULL ...", "DB_Error", TRUE)
#4 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/pear/db/DB/common.php(1928): PEAR->__call("raiseError", (Array:7))
#5 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/pear/db/DB/mysqli.php(936): DB_common->raiseError(-1, NULL, NULL, "ALTER TABLE  `civicrm_grant_app_page` ADD `is_draft` TINYINT(4) DEFAULT NULL ...", "1060 ** Duplicate column name 'is_draft'")
#6 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/pear/db/DB/mysqli.php(406): DB_mysqli->mysqliRaiseError()
#7 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/pear/db/DB/common.php(1234): DB_mysqli->simpleQuery("ALTER TABLE  `civicrm_grant_app_page` ADD `is_draft` TINYINT(4) DEFAULT NULL ...")
#8 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/civicrm/civicrm-core/CRM/Utils/File.php(348): DB_common->query("ALTER TABLE `civicrm_grant_app_page` ADD `is_draft` TINYINT(4) DEFAULT NULL ...")
#9 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/civicrm/civicrm-core/CRM/Utils/File.php(310): CRM_Utils_File::runSqlQuerycoronaun_mumreg3?new_li...", "ALTER TABLE  `civicrm_grant_app_page` ADD `is_draft` TINYINT(4) DEFAULT NULL ...", NULL, TRUE)
#10 /home/coronaun/myukmanifesto.uk/workflow/registry3/sites/default/files/civicrm/ext/biz.jmaconsulting.grantapplications/CRM/Grantapplications/Upgrader/Base.php(128): CRM_Utils_File::s?new_lin...", "/home/coronaun/myukmanifesto.uk/workflow/registry3/sites/default/files/civicr...")
#11 /home/coronaun/myukmanifesto.uk/workflow/registry3/sites/default/files/civicrm/ext/biz.jmaconsulting.grantapplications/CRM/Grantapplications/Upgrader.php(62): CRM_Grantapplications_Upgrader_Base->executeSqlFile("sql/upgrade_4301.sql")
#12 /home/coronaun/myukmanifesto.uk/workflow/registry3/sites/default/files/civicrm/ext/biz.jmaconsulting.grantapplications/CRM/Grantapplications/Upgrader/Base.php(75): CRM_Grantapplications_Upgrader->upgrade_4301()
#13 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/civicrm/civicrm-core/CRM/Queue/Task.php(74): CRM_Grantapplications_Upgrader_Base::_queueAdapter(Object(CRM_Queue_TaskContext), "upgrade_4301")
#14 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/civicrm/civicrm-core/CRM/Queue/Runner.php(202): CRM_Queue_Task->run(Object(CRM_Queue_TaskContext))
#15 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/civicrm/civicrm-core/CRM/Queue/Page/AJAX.php(36): CRM_Queue_Runner->runNext(TRUE)
#16 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/civicrm/civicrm-core/CRM/Queue/ErrorPolicy.php(90): CRM_Queue_Page_AJAX::{closure}()
#17 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/civicrm/civicrm-core/CRM/Queue/Page/AJAX.php(38): CRM_Queue_ErrorPolicy->call(Object(Closure))
#18 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/civicrm/civicrm-core/CRM/Core/Invoke.php(278): CRM_Queue_Page_AJAX::runNext()
#19 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/civicrm/civicrm-core/CRM/Core/Invoke.php(68): CRM_Core_Invoke::runItem((Array:13))
#20 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/civicrm/civicrm-core/CRM/Core/Invoke.php(36): CRM_Core_Invoke::_invoke((Array:4))
#21 /home/coronaun/myukmanifesto.uk/workflow/registry3/modules/contrib/civicrm/src/Civicrm.php(88): CRM_Core_Invoke::invoke((Array:4))
#22 /home/coronaun/myukmanifesto.uk/workflow/registry3/modules/contrib/civicrm/src/Controller/CivicrmController.php(80): Drupal\civicrm\Civicrm->invoke((Array:4))
#23 [internal function](): Drupal\civicrm\Controller\CivicrmController->main((Array:4), "queue:ajax:runNext")
#24 /home/coronaun/myukmanifesto.uk/workflow/registry3/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php(123): call_user_func_array((Array:2), (Array:2))
#25 /home/coronaun/myukmanifesto.uk/workflow/registry3/core/lib/Drupal/Core/Render/Renderer.php(573): Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}()
#26 /home/coronaun/myukmanifesto.uk/workflow/registry3/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php(124): Drupal\Core\Render\Renderer->executeInRenderContext(Object(Drupal\Core\Render\RenderContext), Object(Closure))
#27 /home/coronaun/myukmanifesto.uk/workflow/registry3/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php(97): Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->wrapControllerExecutionInRenderContext((Array:2), (Array:2))
#28 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/symfony/http-kernel/HttpKernel.php(158): Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}()
#29 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/symfony/http-kernel/HttpKernel.php(80): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1)
#30 /home/coronaun/myukmanifesto.uk/workflow/registry3/core/lib/Drupal/Core/StackMiddleware/Session.php(57): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#31 /home/coronaun/myukmanifesto.uk/workflow/registry3/core/lib/Drupal/Core/StackMiddleware/KernelPreHandle.php(47): Drupal\Core\StackMiddleware\Session->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#32 /home/coronaun/myukmanifesto.uk/workflow/registry3/core/modules/page_cache/src/StackMiddleware/PageCache.php(106): Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#33 /home/coronaun/myukmanifesto.uk/workflow/registry3/core/modules/page_cache/src/StackMiddleware/PageCache.php(85): Drupal\page_cache\StackMiddleware\PageCache->pass(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#34 /home/coronaun/myukmanifesto.uk/workflow/registry3/core/lib/Drupal/Core/StackMiddleware/ReverseProxyMiddleware.php(47): Drupal\page_cache\StackMiddleware\PageCache->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#35 /home/coronaun/myukmanifesto.uk/workflow/registry3/core/lib/Drupal/Core/StackMiddleware/NegotiationMiddleware.php(52): Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#36 /home/coronaun/myukmanifesto.uk/workflow/registry3/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#37 /home/coronaun/myukmanifesto.uk/workflow/registry3/core/lib/Drupal/Core/DrupalKernel.php(706): Stack\StackedHttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, TRUE)
#38 /home/coronaun/myukmanifesto.uk/workflow/registry3/index.php(19): Drupal\Core\DrupalKernel->handle(Object(Symfony\Component\HttpFoundation\Request))
#39 {main}`

**Comment by cmastro** [comment]
Hi!

I have a similar issue and I have solved doing this:

- Go to "/home/coronaun/myukmanifesto.uk/workflow/registry3/sites/default/files/civicrm/ext/biz.jmaconsulting.grantapplications/sql"
- Verify the content of "upgrade_4301.sql" 
- On my issue, the SQL try to drop an empty table, so I comment the line with the issue, and run again the Update. In your case, comment or edit the line with your issue, and try to run again the Update.

Hope this solve your issue.

**Comment by JoeMurray** [comment]
Thanks for the report and diagnosis. @monishdeb please fix

**Comment by webmaster-cses-org-uk** [comment]
Can confirm that these issues occur on Joomla as well (CiviCRM 5.37.0, Grant Application Pages 1.7.3).

When installing or enabling, site crashes with an HTTP 500 error, looking in Apache error log reveals the following:
`PHP Fatal error:  Cannot declare class CRM_Grant_Info, because the name is already in use in [...]/public_html/media/civicrm/ext/biz.jmaconsulting.grantapplications/CRM/Grant/Info.php on line 0, referer: [...]/?option=com_civicrm&task=civicrm/admin/extensions&action=enable&id=biz.jmaconsulting.grantapplications&key=biz.jmaconsulting.grantapplications`

Also the operations in upgrade_4700.sql fail because the columns are not previously defined in the table in auto_install.sql:
```sql
ALTER TABLE `civicrm_grant_app_page` DROP for_organization;
ALTER TABLE `civicrm_grant_app_page` DROP is_for_organization;
```

I was able to continue by commenting these out.

**Comment by guyiac** [comment]
We're having trouble on install for WordPress and Drupal sites as well. Some look related to the organization issue, another looks like it's related to the field to indicate whether or not the grant page is a draft. Any thoughts on estimates to fix this Joe? I have a client that needs it and might be able to get some funding.

**Comment by JoeMurray** [comment]
A question for @guyiac @greenlanegreb and @webmaster-cses-org-uk : what version of the extension did you install? I see that we have been remiss in not tagging new releases. There were a couple of branches created for CiviCRM v5.x but no tags. Master branch is considerably ahead.

@guyiac thanks very much for the offer of assistance. We're going to put in some effort in the coming week or two to address outstanding issues and tag a new release that installs with WP and Drupal. There is some current work to move CiviGrant from being a core component to a core extension that I have supported for years and Coleman recently did in core. We'll check again to make sure that this extension works with both the Component and core extension.

@guyiac I'll let you know how much time we have spent on this and will greatly appreciate any contribution you can make on behalf of your clients.

**Comment by guyiac** [comment]
I installed 1.7.3 through the Civi UI. Should I try uninstalling and instead installing from the master?

**Comment by JoeMurray** [comment]
Yes.

Joe Murray, PhD (he/him)
President, JMA Consulting
416.466.1281

**Comment by guyiac** [comment]
I installed the master on a Drupal site (5.36.1) that had never had the extension, and it installed without any difficulty. However, it didn't indicate that there were any database updates that needed to be done - I'm not sure if that's expected behavior. Also, I can add Grant fields to a profile, but the New Grant Application Page link gives me a redirect error (no other pages on the site do).

**Comment by webmaster-cses-org-uk** [comment]
I installed 1.7.3 directly through the CiviCRM extensions interface.

![image](https://user-images.githubusercontent.com/72983627/144122486-06e62c5a-6979-46bb-b8f3-742e96184558.png)

Hope that helps?

**Comment by monishdeb** [comment]
I tested against the latest released version 1.8 (mentioned [here](https://civicrm.org/extensions/grant-application-pages) without any issue. @webmaster-cses-org-uk @guyiac please reopen the ticket - https://lab.civicrm.org/extensions/grantapplications/-/issues/168 if this issue occur
